### ISSUE TITLE
global query parameter suggestion

### ISSUE THREAD

**Issue by Guria** [cerebral/cerebral-router#58]
There is use case when query param is used across the app, like current language: `?lang=ru`. It could be solved with mapping such params to some state path directly.
```js
Router(controller, routes, {
  globalParams: {
    lang: ['lang']
  }
})
```
Then router would exclude `globalParams` from passing as signal payload and would add it while stringify. Most obvious drawbacks:
* this params would always present in url
* router bounded signals would not able to receive same named params as payload

cc @christianalfoni

**Comment by christianalfoni** [comment]
Ah, yes, this is interesting!

Hm, a different perspective would be to use signal options for this:

```js
controller.signal('somethingHappened', chain, {
  defaultInput: {
    lang: ['lang']
  }
})
```

So you would use signal factories to give a bit more flexibility. I like your solution better in terms of "always being there", but if there are scenarios where that is not the case, it is not too much trouble creating a factory instead for signals where you set the option above.

Hm, maybe we should try to list up some scenarios before making a decision here

**Comment by awei01** [comment]
Not sure if this is the exact same issue, but I have a situation where I want to use the query param to trigger a modal window that may not be specific to a certain route; the modal could appear on the home route, or on a nested route. 

Currently, I can rely on the this module to catch addressbar.onChange event and simply create an <a href="?module=FooModule">open foo module</a>. But, it would be nicer if I could keep to the same coding pattern and pass the responsibility over to `cerebral-module-router`

**Comment by christianalfoni** [comment]
Though what if we make a more explicit separation of what a **param** and a **query** is. With the Cerebral router params of course changes across signals/urls because they are part of the url, but queries do not change. Queries are consistent. What you rather do is:

```js
function setLanguage({services}) {
  services.router.updateQuery({
    lang: 'EN'
  })
}
```

You can also clear the query:

```js
function setLanguage({services}) {
  services.router.clearQuery()
}
```

So the query will always be there, also when you trigger new routes/signals. We say: "THE QUERY IS GLOBAL".

**Comment by Guria** [comment]
And what should happen when `/someRoute?lang=EN` is triggered?
'someRoute' will be mapped to someSignal. But where goes `{lang: 'EN'}` part? Who is responsible for it?

**Comment by christianalfoni** [comment]
Lets iterate a bit here :-)

Okay, so I think it is a better idea to keep the queries away from the signal payloads.

```js
services.router.getQuery() // {foo: 'bar'}
services.router.setQuery() // Clears current and sets new
services.router.updateQuery() // Merges with current query
services.router.clearQuery() // Removes query
```

If `/someRoute?lang=EN` fires any action can grab that information explicitly with `services.router.getQuery('lang')`

If I for example change some filters, page etc. I would update the query manually in some action:

```js
function changePage({input, state, services) {
  services.router.updateQuery({page: input.page})
  state.set('app.currentPage', input.page)
}
```

So queries follows along when you trigger signals/urls. That means when you trigger a signal related to a url `signals.thisHappened({foo: 'bar'})` it will just be part of the payload to the signal. To actually update the query you have to do that in an action: `services.router.updateQuery({foo: input.foo})`.

Handling queries as something global and through the service allows us to solve anything it seems to me. We can later see if there are ways we can make f.ex. operators to do this:

```js
import {updateQuery} from 'cerebral-module-router`

updateQuery('input:foo', 'foo')
```

Type of thing

**Comment by edgesoft** [comment]
Don't have a strong opinion on this. Either way! Haven't had a use case for this with Cerebral yet so hard to say anything about it.

**Comment by christianalfoni** [comment]
Okay, so I have a different proposition to handle global queries.

Instead of using a service we give the router a namespace:

```js
{
  router: {
    query: {}
  }
}
```

This query is synced with the url at all times. So a `change` event in Cerebral will make the router resync the url with whatever is in the query.

So typically when a page is loaded, whatever is on the query will be synced into the store, for example `/?lang=EN`

```js
{
  router: {
    query: {
      lang: 'EN'
    }
  }
}
```

If an action changes the language it can, if it wants to do: `state.set('router.query.lang', 'NO')` and when Cerebral runs the `change` event the query is updated.

If you want to clean out queries you typically do that on a component unmount or a routing/mounted of something else. Just using `state.unset('router.query.something')` is enough.

Now we have state driven queries :)

**Comment by Guria** [comment]
@christianalfoni Yeah, I have though about something similar. But there is drawbacks with this case too.
First of all, how do you suggest router to change a state? We are not injecting into signals anymore.

**Comment by Guria** [comment]
@christianalfoni Yeah, I have though about something similar. But there is drawbacks with this case too.
It seems you forgot that router must update state from url as well. We could use `module.addState` to set initial state from url, but we wait `modulesLoaded` event to actual parse of url so it doesn't fit. And also we don't have a way to set state from later (browser back button which change query param) because it allowed only within signal.

**Comment by christianalfoni** [comment]
Yeah, I got to this conclusion when I got to bed yesterday and then got up because I could not sleep... thought about it some more. But yeah, that is a problem :)

The idea of having it state driven though I like. It just fits with the rest of it.

Just to play around with the idea some more. We could still do it like this, but with different behaviour.

1. The page loads and queries are passed in as payload to signals as it is now. For example:

`/?lang=EN` => `{lang: 'EN'}`

2. To actually synchronise the query you have to put it into: `state.set('router.query.lang', input.lang)`

3. Now the router will keep this in the url query whenever you do a route change

It would allow you to create factories like: `homeRouted: setQuery(openHome)`. This is also a security concern. We do not really want to automatically synchronize queries with the state store automatically from the URL, the reason being that you might be putting stuff there that you should not. Now we give the developer manual control of it, but with a synchronize ability by putting it into the routers namespace.

**Comment by christianalfoni** [comment]
Yeah, I got to this conclusion when I got to bed yesterday and then got up because I could not sleep... thought about it some more. But yeah, that is a problem :)

Hm, I think we need to break this up.

- Now we are passing queries as payload to the signal. This can be a problem with global queries because this payload will be passed into any url bound signal implicitly. This could even be a situation now. I could add: `?` to any Cerebral page and start overriding payloads to signals

- Some queries we want to be global, like `lang`, but other we want to be local to the signal, like `limit`

SO!

1. We should remove passing queries as payload to signals as it is a "point of attack". This needs to be controlled by the developer

2. We need a concept of defining global and local queries

SO!

1. Queries should not have any meaning without the developer explicitly grabbing the query and doing something with it. To get this functionality we have to use a service I think:

`services.router.getQuery('lang')`

2. There is also a different problem with having queries state driven and that is to remove them on route change. Local queries will not work as we can not do an `unset` outside a signal. Which means we have to use a service here as well:

`services.router.setGlobalQuery('lang')` or `services.router.setQuery('limit')` (removed on route change)

Queries are a lot more complex than we give it credit for I think. So making it state driven is a security risk and it will not allow us the flexibility we need. I would love to have it state driven, but the urls are not state driven either. They do not exist in the tree and you do not change the state to change the url. I think the nature of urls is too complex to make it state driven.

**Comment by Guria** [comment]
> We should remove passing queries as payload to signals as it is a "point of attack". This needs to be controlled by the developer

* it is [not possible to override param defined as dynamic part](https://tonicdev.com/57a9d54d3a22f913005c0e44/57b2fa2b73c03f1200296d42)
* it is a client side javascript. you can't prevent user to call signal with any paramater he want
* only issue can be passing extra params that signal doesn't use

> We need a concept of defining global and local queries

Agreed. But I don't want to do it through service calls. It makes query string to be alternate source of state and ties app to router.

My thought was to extend route syntax with query definitions:
```js
{
  '/:dynamicParam?queryParam1&queryParam2': 'smthHappened'
}
```
It would mean that only `queryParam1` and `queryParam2` from query would be passed to `smthHappened` signal. But need to think of another syntax because it confilcts with optional dynamic param. Would be great to sync rest of params with `{router:{query:{}}}` in state, but I don't know a good way to do it right now.

**Comment by christianalfoni** [comment]
What I mean be "point of attack" is:

Somebody shares a bad url to the app: `/something?value=someBadScript`. If the signal expects to get `value` when signal triggers manually (not from url) it will think it actually got the value, but it was not from the app, but the url. Then it is put into state tree and if the hack is successful the client triggers a script based on a url. It is an extreme edge case, but possible :)

But we have to answer a question first. Can queries be state driven? I do not think they can: because

1. We tried to make the url state driven, but we figured out it was impossible because of `back` button and some other things I can not remember

2. Queries are not necessarily bound to a specific url. They can be bound globally. So `/home?lang=EN` and `/admin?lang=EN` are both valid, but `/users?limit=20` is only valid on `/users`

These two facts says something about the complexity of routing I think. And again:

1. It is currently impossible to set queries into state tree based on a url. We can not hijack a signal. And if we did it outside of a signal the debugger would not know about it

2. When a url changes we have to remove queries based on them being bound globally or to the current url, which we can not do either because we have to hijack the signal the same way as above

And the state of queries is always seen in the addressbar, no matter how we update it :) So I do not think it will cause any confusion if done through a service. And debugger shows that you are using the service doing the updates to the query so easy to debug.

Anyways, just pushing from the other side :)
